contracts/liquid-locking/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+#![no_std]
+
+multiversx_sc::imports!();
+
+#[multiversx_sc::contract]
+pub trait LiquidLocking {
+    #[init]
+    fn init(&self, unbond_period: u64) {
+        self.unbond_period().set_if_empty(unbond_period);
+    }
+
+    #[only_owner]
+    #[endpoint]
+    fn set_unbond_period(&self, unbond_period: u64) {
+        self.unbond_period().set(unbond_period);
+    }
+
+    #[only_owner]
+    #[endpoint]
+    fn whitelist_token(&self, token: TokenIdentifier) {
+        require!(token.is_valid_esdt_identifier(), "invalid token provided");
+        let _ = self.token_whitelist().insert(token);
+    }
+
+    #[only_owner]
+    #[endpoint]
+    fn blacklist_token(&self, token: &TokenIdentifier) {
+        let _ = self.token_whitelist().swap_remove(token);
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn stake(&self) {
+        let payments = self.call_value().all_esdt_transfers();
+        let caller = self.blockchain().get_caller();
+        for payment in payments.iter() {
+            self.validate_payment(&payment);
+            self.stake_token(&caller, payment);
+        }
+    }
+
+    fn validate_payment(&self, payment: &EsdtTokenPayment) {
+        require!(payment.token_nonce != 0, "invalid token provided");
+        require!(payment.amount != 0, "amount must be greater than 0");
+        require!(
+            self.token_whitelist().contains(&payment.token_identifier),
+            "token is not whitelisted"
+        );
+    }
+
+    fn stake_token(&self, caller: &ManagedAddress, payment: EsdtTokenPayment) {
+        self.staked_tokens(caller, &payment.token_identifier)
+            .update(|amount| *amount += &payment.amount);
+    }
+
+    #[endpoint]
+    fn unstake(&self, tokens: ManagedVec<EsdtTokenPayment<Self::Api>>) {
+        let caller = self.blockchain().get_caller();
+        let block_epoch = self.blockchain().get_block_epoch();
+        for token in tokens.iter() {
+            self.staked_tokens(&caller, &token.token_identifier)
+                .update(|staked_amount| {
+                    if *staked_amount >= token.amount {
```

Comment:
I'm not really a fan of "hiding" the fact that a user had a wrong input. I would actually throw an error and fail the whole transaction. Or at least emit an "error log" for the else case :)

Reply:
Can we also refactor a bit the logic in the lambda (pull it in some functions) to be easier to read? The same for unbond below

Reply:
Also, like @claudiulataretu mentioned on slack, some view functions would be great (what tokens and how much users have staked, what tokens are whitelisted, unstaked values and unbond periods come to mind)